viewing_party/party.py

```diff
@@ -1,23 +1,196 @@
+# return movie user has watched but friends haven't as a list of dictionaries
+# compare {} in "watched" and "friends":"watched", return the unique one in "watched".
+
+# make a list of friends wathced movie titles. Set makes it more efficient.
+    friends_movies = set()
+    for movie_dict in user_data["friends"]:
+        for movie in movie_dict["watched"]:
+            friends_movies.add(movie["title"])
+
+# loop throught user watched list, return ones different than friends watched list.
+    user_unique = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_movies:
+            user_unique.append(movie)
+
+    return user_unique       
+
+def get_friends_unique_watched(user_data):
+# return movie at least one of the friends have watched, but user hasn't as a list of dict
+    user_movies = set()
+    for movie in user_data["watched"]:
+        user_movies.add(movie["title"])
 
+    friends_unique = []
+    for movie_dict in user_data["friends"]:
+        for movie in movie_dict["watched"]:
+            if movie["title"] not in user_movies and movie not in friends_unique:
+                friends_unique.append(movie)
+
+    return friends_unique
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    
+    recommendations = []
+
+    if not user_data:
+        return recommendations
+    
+    user_movies = []
+    for u_movies in user_data["watched"]:
+        user_movies.append(u_movies["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if (movie["title"] not in user_movies and
+                movie["host"] in user_data["subscriptions"] and
+                movie not in recommendations):
+                recommendations.append(movie)
+
+    return recommendations
+
+
+#make a list of recommended movies
+    #add only if:
+    #not in user_data watchlist
+    #in friends dictionary
+    #in user's subsciptions
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+# rec_genre = [], not in "watched", in at least one friends "watched", "genre" is user most frequent
+# find genre
+    fav_genre = get_most_watched_genre(user_data)
+    friends_list = get_friends_unique_watched(user_data)
+
+    rec_by_genre = []
+    for movie in friends_list:
+        if movie["genre"]  == fav_genre:
+            rec_by_genre.append(movie)
+
+    return rec_by_genre
+
+def get_rec_from_favorites(user_data):
+# in "favorites", not in friends "watched"
+    unique_movies = get_unique_watched(user_data)
+    rec_from_fav = []
+
+    for movie in user_data["favorites"]:
+        for unique_movie in unique_movies:
+            if movie == unique_movie:
+                rec_from_fav.append(movie)
+
+    return rec_from_fav

```

Comment:
Same as before, we could make this a list comprehension as well! Feel free to try it out if you would like!